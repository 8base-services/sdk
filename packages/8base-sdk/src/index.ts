export * from '@8base/apollo-client';
export * from '@8base/apollo-links';
export * from '@8base/auth';
export * from '@8base/utils';
export * from '@8base/validate';
