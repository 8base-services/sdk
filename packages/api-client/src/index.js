export { importTables } from './importTables';
export { importData } from './importData';
export { exportTables } from './exportTables';
export { Client } from './Client';
