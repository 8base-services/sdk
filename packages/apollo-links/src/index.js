export { TokenRefreshLink } from './TokenRefreshLink';
export { fileUploadLink } from './fileUploadLink';
export { createFile } from './utils/createFile';
