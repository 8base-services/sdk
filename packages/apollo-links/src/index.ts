export { TokenRefreshLink } from './TokenRefreshLink';
export { AuthLink } from './AuthLink';
export { AuthHeadersLink } from './AuthHeadersLink';
export { SuccessLink } from './SuccessLink';
export { SignUpLink } from './SignUpLink';
export * from './types';

