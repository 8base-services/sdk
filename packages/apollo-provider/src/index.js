// @flow

export { ApolloProvider } from './ApolloProvider';
