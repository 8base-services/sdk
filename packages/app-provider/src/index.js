export { EightBaseAppProvider } from './EightBaseAppProvider';
export { FragmentsSchemaContainer } from './FragmentsSchemaContainer';
