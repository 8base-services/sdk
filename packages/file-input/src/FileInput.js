// @flow
import React from 'react';
import { withApollo } from 'react-apollo';
import * as filestack from 'filestack-js';
import gql from 'graphql-tag';

type FileValue = {
  fileId: string,
  filename: string,
  id?: string,
  downloadUrl?: string,
};

type FileInputValue = FileValue | FileValue[];

type FileInputProps = {
  client: Object,
  onChange: (value: FileInputValue) => void,
  children: ({ pick: () => Promise<void>, value: ?FileInputValue, error: ?Object }) => React$Node,
  value?: FileInputValue,
  maxFiles?: number,
  onUploadFinish?: (value: FileInputValue) => Promise<FileInputValue>,
};

type FileInputState = {
  path: ?string,
  error: ?Object,
};

const FILE_UPLOAD_INFO_QUERY = gql`
  query FileUploadInfo {
    fileUploadInfo {
      policy
      signature
      apiKey
      path
    }
  }
`;

class FileInput extends React.Component<FileInputProps, FileInputState> {
  filestack: Object;
  filestackPromise: Promise<void>;

  static defaultProps = {
    maxFiles: 1,
  };

  constructor(props) {
    super(props);

    this.state = {
      path: null,
      error: null,
    };
  }

  componentDidMount() {
    this.filestackPromise = this.initFilestack();
  }

  async initFilestack() {
    const { client } = this.props;

    let response = null;

    try {
      response = await client.query({ query: FILE_UPLOAD_INFO_QUERY });
    } catch (e) {
      this.setState({ error: e });

      return;
    }

    const { apiKey, policy, signature, path } = response.data.fileUploadInfo;

    this.setState({ path });

    this.filestack = filestack.init(apiKey, {
      security: {
        policy,
        signature,
      },
    });
  }

  onUploadDone = async ({ filesUploaded }) => {
    let value = filesUploaded.map(({ handle, filename }) => ({
      fileId: handle,
      filename,
    }));

    const { maxFiles, onUploadFinish } = this.props;

    if (maxFiles === 1) {
      value = value[0];
    }

    if (typeof onUploadFinish === 'function') {
      value = await onUploadFinish(value);
    }

    this.props.onChange(value);
  };

  collectPickerOptions = () => {
    const { maxFiles } = this.props;
    const { path } = this.state;

    return {
      onUploadDone: this.onUploadDone,
      storeTo: {
        path,
      },
      maxFiles,
    };
  };

  pick = async () => {
    await this.filestackPromise;

    const pickerOptions = this.collectPickerOptions();

    this.filestack.picker(pickerOptions).open();
  };

  render() {
    const { children, value } = this.props;

    const { error } = this.state;

    return children({ pick: this.pick, value, error });
  }
}

FileInput = withApollo(FileInput);

export { FileInput };
