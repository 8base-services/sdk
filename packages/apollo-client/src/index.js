// @flow

import gql from 'graphql-tag';
import { type ApolloClientOptions, ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloLink } from 'apollo-link';
import { BatchHttpLink } from 'apollo-link-batch-http';
import { createAuthLink, fileUploadLink, SubscriptionLink } from '@8base/apollo-links';

type EightBaseApolloClientOptions = {
  uri: string,
  getAuthState: Function,
  getRefreshTokenParameters: Function,
  onAuthSuccess: Function,
  onAuthError?: Function,
  onIdTokenExpired?: Function,
} & ApolloClientOptions;

/**
 * Extended Apollo Client by 8base several links.
 *
 * @param {Object} config - The Apollo Client config.
 * @param {string} config.uri Endpoint of the GraphQl server.
 * @param {Function} config.getAuthState - The function which are using to get auth state.
 * @param {Function} config.getRefreshTokenParameters - The function which are using for get refresh token parameters.
 * @param {Function} config.onAuthSuccess - The callback which called when attempt to refresh authentication is success.
 * @param {Function} [config.onAuthError] - The callback which called when attempt to refresh authentication is failed.
 * @param {Function} [config.onIdTokenExpired] - The callback which called when id token is expired.
 *
 * @return instance of the Apollo Client
 */
class EightBaseApolloClient extends ApolloClient {
  constructor(config: EightBaseApolloClientOptions) {
    const {
      uri,
      getAuthState,
      getRefreshTokenParameters,
      onAuthSuccess,
      onAuthError,
      onIdTokenExpired,
    } = config;

    let { cache } = config;

    if (!cache) {
      cache = new InMemoryCache();
    }

    const authLink = createAuthLink({
      getAuthState,
      getRefreshTokenParameters,
      onAuthSuccess,
      onAuthError,
      onIdTokenExpired,
    });

    const subscriptionLink = new SubscriptionLink();

    const batchHttpLink = new BatchHttpLink({ uri });

    const link = ApolloLink.from([
      fileUploadLink,
      authLink,
      subscriptionLink,
      batchHttpLink,
    ]);

    super({ cache, link });
  }
}

export { EightBaseApolloClient, gql, InMemoryCache };
