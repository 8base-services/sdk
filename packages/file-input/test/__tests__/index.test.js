import React from 'react';
import renderer from 'react-test-renderer';

import { FileInput } from '../../src';

let mock_onUploadDone = null;
let mock_client = null;

jest.mock('react-apollo', () => {
  const client = {
    query: jest.fn(() => {
      mock_client = client;

      return {
        data: {
          fileUploadInfo: {
            apikey: 'apikey',
            policy: 'policy',
            signature: 'signature',
            path: 'path',
          },
        },
      };
    }),
  };

  return {
    withApollo: (BaseComponent) => (props) => <BaseComponent { ...props } client={ client } />,
  };
});

jest.mock('filestack-js', () => {
  const picker = jest.fn(({ onUploadDone }) => ({
    open: () => {
      mock_onUploadDone = onUploadDone;
    },
  }));

  return {
    mock_picker: picker,
    init: () => ({
      picker,
      session: {},
    }),
  };
});

const filestack = require('filestack-js');

beforeEach(() => {
  jest.clearAllMocks();
});

describe('should call onChange when file is uploaded', () => {
  it('for single file input', async () => {
    const renderFileInputView = jest.fn(() => null);
    const onChange = jest.fn();

    renderer.create(
      <FileInput onChange={ onChange }>
        { renderFileInputView }
      </FileInput>,
    );

    expect(renderFileInputView).toHaveBeenCalledTimes(1);

    await renderFileInputView.mock.calls[0][0].pick();

    expect(mock_client.query).toHaveBeenCalledTimes(1);
    expect(mock_client.query.mock.calls[0]).toMatchSnapshot();

    const originalFile = new File([''], 'filename');

    await mock_onUploadDone({ filesUploaded: [{ handle: 'handle', filename: 'filename', originalFile }] });

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange.mock.calls[0][0]).toEqual({ fileId: 'handle', filename: 'filename', public: false });
    expect(onChange.mock.calls[0][1]).toEqual(originalFile);
  });

  it('for single file input with public modifier', async() => {
    const renderFileInputView = jest.fn(() => null);
    const onChange = jest.fn();

    renderer.create(
      <FileInput onChange={ onChange } public>
        { renderFileInputView }
      </FileInput>,
    );

    expect(renderFileInputView).toHaveBeenCalledTimes(1);

    await renderFileInputView.mock.calls[0][0].pick();

    expect(mock_client.query).toHaveBeenCalledTimes(1);
    expect(mock_client.query.mock.calls[0]).toMatchSnapshot();

    const originalFile = new File([''], 'filename');

    await mock_onUploadDone({ filesUploaded: [{ handle: 'handle', filename: 'filename', originalFile }] });

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange.mock.calls[0][0]).toEqual({ fileId: 'handle', filename: 'filename', public: true });
    expect(onChange.mock.calls[0][1]).toEqual(originalFile);
  });

  it('for multiple files input', async() => {
    const renderFileInputView = jest.fn(() => null);
    const onChange = jest.fn();

    renderer.create(
      <FileInput onChange={ onChange } maxFiles={ 3 }>
        { renderFileInputView }
      </FileInput>,
    );

    expect(renderFileInputView).toHaveBeenCalledTimes(1);

    await renderFileInputView.mock.calls[0][0].pick();

    expect(mock_client.query).toHaveBeenCalledTimes(1);
    expect(mock_client.query.mock.calls[0]).toMatchSnapshot();

    const originalFile1 = new File([''], 'filename1');
    const originalFile2 = new File([''], 'filename2');
    const originalFile3 = new File([''], 'filename3');

    await mock_onUploadDone({
      filesUploaded: [
        { handle: 'handle1', filename: 'filename1', originalFile: originalFile1 },
        { handle: 'handle2', filename: 'filename2', originalFile: originalFile2 },
        { handle: 'handle3', filename: 'filename3', originalFile: originalFile3 },
      ],
    });

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange.mock.calls[0][0]).toEqual(
      [
        { fileId: 'handle1', filename: 'filename1', public: false },
        { fileId: 'handle2', filename: 'filename2', public: false },
        { fileId: 'handle3', filename: 'filename3', public: false },
      ],
    );
    expect(onChange.mock.calls[0][1]).toEqual(
      [
        originalFile1,
        originalFile2,
        originalFile3,
      ],
    );
  });

  it('for input with custom on upload handler', async() => {
    const renderFileInputView = jest.fn(() => null);
    const onChange = jest.fn();

    const onUploadDone = (value) => new Promise((resolve) => {
      setTimeout(() => {
        resolve({ ...value, id: 'id', downloadUrl: 'downloadUrl' });
      }, 50);
    });

    renderer.create(
      <FileInput onChange={ onChange } onUploadDone={ onUploadDone }>
        { renderFileInputView }
      </FileInput>,
    );

    expect(renderFileInputView).toHaveBeenCalledTimes(1);

    await renderFileInputView.mock.calls[0][0].pick();

    expect(mock_client.query).toHaveBeenCalledTimes(1);
    expect(mock_client.query.mock.calls[0]).toMatchSnapshot();

    await mock_onUploadDone({ filesUploaded: [{ handle: 'handle', filename: 'filename' }] });

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange.mock.calls[0][0]).toEqual({ fileId: 'handle', filename: 'filename', id: 'id', downloadUrl: 'downloadUrl', public: false });
  });
});

describe('FileInput', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('allows to pass custom options to pick method', async () => {
    const renderFileInputView = jest.fn(() => null);

    renderer.create(
      <FileInput>
        { renderFileInputView }
      </FileInput>,
    );

    await renderFileInputView.mock.calls[0][0].pick({
      accept: ['image/*'],
      fromSources: ['dropbox'],
    });

    expect(filestack.mock_picker).toHaveBeenCalledWith({
      exposeOriginalFile: true,
      onUploadDone: expect.any(Function),
      storeTo: {
        path: 'path',
      },
      maxFiles: 1,
      accept: ['image/*'],
      fromSources: ['dropbox'],
    });
  });

  it('doesn\'t allow to rewrite maxFiles and onUploadDone options', async () => {
    const renderFileInputView = jest.fn(() => null);

    renderer.create(
      <FileInput maxFiles={ 3 }>
        { renderFileInputView }
      </FileInput>,
    );

    await renderFileInputView.mock.calls[0][0].pick({
      accept: ['image/*'],
      fromSources: ['dropbox'],
      maxFiles: 2,
      onUploadDone: 'not function',
    });

    expect(filestack.mock_picker).toHaveBeenCalledWith({
      exposeOriginalFile: true,
      onUploadDone: expect.any(Function),
      storeTo: {
        path: 'path',
      },
      maxFiles: 3,
      accept: ['image/*'],
      fromSources: ['dropbox'],
    });
  });
});

