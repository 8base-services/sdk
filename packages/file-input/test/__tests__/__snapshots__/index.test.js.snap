// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should call onChange when file is uploaded 1`] = `
Array [
  Object {
    "query": Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "FileUploadInfo",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "fileUploadInfo",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "policy",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "signature",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "apiKey",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "path",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 110,
        "start": 0,
      },
    },
  },
]
`;
